### PR TITLE
Add 'save_format' param to Keras autolog()

### LINKED ISSUE #7024: [FR] Allowing changing the Keras serialization format when using `mlflow.keras.autolog()`
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

Currently, it is not possible to change the serialization format of the Keras models (tf) to HDF% when using the `mlflow.keras.autolog()`. It would be great if this could be configured. One way could be for `autolog()` to take a `kwargs` that is passed to inner functions so that serialization configuration can be set when using `autolog()` 

### Motivation

> #### What is the use case for this feature?

> #### Why is this use case valuable to support for MLflow users in general?

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by balvisio**
Signed-off-by: Bruno Alvisio <bruno.alvisio@gmail.com>

## Related Issues/PRs

Resolve https://github.com/mlflow/mlflow/issues/7024

## What changes are proposed in this pull request?

Allow configuring the format in which the model is saved in `keras.autolog()` 

## How is this patch tested?

Added test to load model in 'HDF5' format

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/keras.py [hunk 1]
```diff
@@ -681,6 +685,12 @@ def autolog(
             stacklevel=2,
         )
 
+    if save_format != "tf" and save_format != "h5":
```
Comment (harupy) [inline mlflow/keras.py:688]:
```suggestion
    if save_format not in ["tf", "h5"]:
```

File: mlflow/keras.py [hunk 2]
```diff
@@ -681,6 +685,12 @@ def autolog(
...
+        raise Exception(
+            "Invalid value for `save_format` argument. "
+            "Valid values for `save_format` are `tf` or `h5` only."
+        )
```
Comment (harupy) [inline mlflow/keras.py:692]:
```suggestion
        raise MlflowException(
            "Invalid value for `save_format` argument. "
            "Valid values for `save_format` are `tf` or `h5` only.",
            INVALID_PARAMETER,
        )
```

File: mlflow/keras.py [hunk 3]
```diff
@@ -733,6 +743,7 @@ def on_train_end(self, logs=None):
                         self.model,
                         artifact_path="model",
                         registered_model_name=registered_model_name,
+                        **{"save_format": save_format},
```
Comment (harupy) [inline mlflow/keras.py:746]:
```suggestion
                        save_format=save_format,
```

File: tests/keras/test_keras_autolog.py [hunk 4]
```diff
@@ -410,3 +410,20 @@ def test_autolog_registering_model(random_train_data, random_one_hot_labels):
...
+def test_autolog_load_saved_hdf5_model(random_train_data, random_one_hot_labels):
+    mlflow.keras.autolog(save_format="h5")
+
+    data = random_train_data
+    labels = random_one_hot_labels
+
+    model = create_model()
+    with mlflow.start_run() as run:
+        model.fit(data, labels, epochs=10)
+        mlflow.keras.load_model("runs:/" + run.info.run_id + "/model")
```
Comment (harupy) [inline tests/keras/test_keras_autolog.py:424]: Is there a way to check the model is saved in the specified format?

File: mlflow/keras.py [hunk 5]
```diff
@@ -681,6 +685,12 @@ def autolog(
             stacklevel=2,
         )
 
+    if save_format != "tf" and save_format != "h5":
+        raise Exception(
+            "Invalid value for `save_format` argument. "
+            "Valid values for `save_format` are `tf` or `h5` only."
```
Comment (harupy) [inline mlflow/keras.py:691]:
```suggestion
            "Valid values for `save_format` are 'tf' or 'h5' only."
```

File: mlflow/keras.py [hunk 6]
```diff
@@ -681,6 +685,12 @@ def autolog(
             stacklevel=2,
         )
 
+    if save_format != "tf" and save_format != "h5":
+        raise Exception(
+            "Invalid value for `save_format` argument. "
```
Comment (harupy) [inline mlflow/keras.py:690]: Can we include the invalid value in the message?

**Comment by harupy** [comment]
@balvisio Thanks for the PR! Left some comments :)

File: mlflow/keras.py (hunk 6 above)
Comment (harupy) [reply mlflow/keras.py:690]:
```suggestion
            "Invalid value for `save_format` argument: {save_format}. "
```

**Comment by balvisio** [comment]
@harupy Sorry for not following up, has been a little bit busy. Thank you so much for the comments. I agree with your suggestions.

**Comment by balvisio** [comment]
@harupy I have noticed that there is a warning message that the Keras support will be removed in a future version of MLFlow and will be replaced for TF. When I looked at the TF-related source code for saving the model looks like it supports "SavedModel" format only and it would be quite a bit a work to support HDF5. Not sure if it is worth the effort to support both formats. What do you think?

**Comment by harupy** [comment]
@balvisio Keras support won't be removed. Sorry for the confusion. In MLflow 2.0.0 (that will be released soon), `mlflow.tensorflow` will support `tf.keras` model. We'll merge this PR first, then export the changes in this PR to `mlflow.tensorflow` before releasing 2.0.0 :)

**Comment by balvisio** [comment]
Got it. Yeah, I got confused because of this warning `Autologging support for keras >= 2.6.0 has been deprecated and will be removed`. Is not applicable then?

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**